### ISSUE TITLE
[API] PVC stuck in pending for NFS/EFS storage class. Corresponding PV is not automatically created.

### ISSUE THREAD

**Issue by nagapavan** [kubernetes/kubernetes#65937]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug
/sig aws

**What happened**:
I have created a PVC for a [EFS volume](https://github.com/kubernetes-incubator/external-storage/tree/master/aws/efs), with ConfigMap in place. 
``` yaml
{
  "kind": "PersistentVolumeClaim",
  "apiVersion": "v1",
  "metadata": {
    "name": "efs",
    "namespace": "default",
    "annotations": {
      "volume.beta.kubernetes.io/storage-class": "aws-efs"
    }
  },
  "spec": {
    "accessModes": [
      "ReadWriteMany"
    ],
    "resources": {
      "requests": {
        "storage": "1Gi"
      }
    }
  }
}
```
`aws-efs` StorageClass is available:
```bash
# kubectl get sc
NAME      PROVISIONER           AGE
aws-efs   example.com/aws-efs   1d
```

However the PVC creation is stuck in state pending.
kubectl status of PVC:
```bash
# kubectl describe pvc efs
Name:          efs
Namespace:     default
StorageClass:  aws-efs
Status:        Pending
Volume:
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-class=aws-efs
               volume.beta.kubernetes.io/storage-provisioner=example.com/aws-efs
Finalizers:    []
Capacity:
Access Modes:
Events:
  Type    Reason                Age                From                         Message
  ----    ------                ----               ----                         -------
  Normal  ExternalProvisioning  1m (x62 over 16m)  persistentvolume-controller  waiting for a volume to be created, either by external provisioner "example.com/aws-efs" or manually created by system administrator
```

**What you expected to happen**:
Corresponding PV should have been automatically created. Since the StorageClass annotation has been used.

**How to reproduce it (as minimally and precisely as possible)**:
Follow steps on [EFS PVC creation](https://github.com/kubernetes-incubator/external-storage/tree/master/aws/efs), with a valid EFS mount details.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

> Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T12:22:21Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
> Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2017-12-15T20:55:30Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): CentOS Linux 7
- Kernel (e.g. `uname -a`): 3.10.0-693.21.1.el7
- Install tools:
- Others:

**Comment by pictolearn** [comment]
I came across the same issue using KOPS.  I have followed whatever was specified here https://medium.com/@while1eq1/using-amazon-efs-in-a-multiaz-kubernetes-setup-57922e032776 but my pod file look different with an nginx instance trying to mount to the EFS cluster

Here is how the nginx yaml looks

`apiVersion: apps/v1 # for versions before 1.9.0 use apps/v1beta2
kind: Deployment
metadata:
  name: nginx-deployment
spec:
  selector:
    matchLabels:
      app: nginx
  replicas: 1 # tells deployment to run 2 pods matching the template
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.7.9
        ports:
        - containerPort: 80
        volumeMounts:
        - name: efs
          mountPath: /usr/share/nginx/html
      volumes:
        - name: efs
          persistentVolumeClaim:
            claimName: efs`

**Comment by tonybranfort** [comment]
If you're using rbac (which it seems kops does by default), make sure you've created a serviceAccount.  Look at the [deployment.yaml](https://github.com/kubernetes-incubator/external-storage/blob/master/aws/efs/deploy/deployment.yaml) rather than the [manifest.yaml](https://github.com/kubernetes-incubator/external-storage/blob/master/aws/efs/deploy/manifest.yaml). 
 
You'll also need [rbac.yaml](https://github.com/kubernetes-incubator/external-storage/blob/master/aws/efs/deploy/rbac.yaml).  The readme describes this but doesn't mention the service account.

**Comment by nagapavan** [comment]
@tonybranfort I was using Service-acccount, rbac and cluster-role binding to the provisioner. There was a bit of challenge in case of cluster-role binding case, I had to fetch existing role binding entry and patch it for each new `namespace`.
Still ran into the issue. :(

**Comment by roboman2444** [comment]
/remove-lifecycle stale
This is still an issue in 1.10.11

**Comment by methuz** [comment]
Somehow EKS does not support ReadWriteMany try ReadWriteOnce

**Comment by saadzaman** [comment]
wondering if you got the solution for this ?

**Comment by amitvashist7** [comment]
@[saadzaman](/saadzaman)

Thank you very much for the suggestion: `ReadWriteOnce` work me as well.

**Comment by divyangjp** [comment]
> @[saadzaman](/saadzaman)
> 
> Thank you very much for the suggestion: `ReadWriteOnce` work me as well.

EFS or NFS should support `ReadWriteMany`. That's the whole point of using them!

**Comment by dragonknight88** [comment]
Following up on @tonybranfort suggestion. It worked for me. 

You need to configure ServiceAccount and ClusterRole. @tonybranfort  provided all the links as well.

**Comment by newalp** [comment]
@tonybranfort provided the right fix.
In summary I had to update my deployment and add ClusterRole, ClusterRoleBinding, Role, RoleBinding, ServiceAccount.
The deployment template provided was expecting a directory on my efs volume called /persistentvolumes. I had to replace this with just / for the pod to start.

See below for configuration I used:

`---
apiVersion: v1
kind: ConfigMap
metadata:
  name: efs-provisioner
data:
  file.system.id: xxxxxxxxxxxxxx
  aws.region: xxxxxxxxxxxxx
  provisioner.name: example.com/aws-efs
  dns.name: ""
---
kind: Deployment
apiVersion: extensions/v1beta1
metadata:
  name: efs-provisioner
spec:
  replicas: 1
  strategy:
    type: Recreate 
  template:
    metadata:
      labels:
        app: efs-provisioner
    spec:
      serviceAccount: efs-provisioner
      containers:
        - name: efs-provisioner
          image: quay.io/external_storage/efs-provisioner:latest
          env:
            - name: FILE_SYSTEM_ID
              valueFrom:
                configMapKeyRef:
                  name: efs-provisioner
                  key: file.system.id
            - name: AWS_REGION
              valueFrom:
                configMapKeyRef:
                  name: efs-provisioner
                  key: aws.region
            - name: DNS_NAME
              valueFrom:
                configMapKeyRef:
                  name: efs-provisioner
                  key: dns.name
                  optional: true
            - name: PROVISIONER_NAME
              valueFrom:
                configMapKeyRef:
                  name: efs-provisioner
                  key: provisioner.name
          volumeMounts:
            - name: pv-volume
              mountPath: /persistentvolumes
      volumes:
        - name: pv-volume
          nfs:
            server: xxxxxxxxxxxxxxx.efs.xxxxxxxxxxx.amazonaws.com
            path: /
---
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: aws-efs
provisioner: example.com/aws-efs
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: efs
  annotations:
    volume.beta.kubernetes.io/storage-class: "aws-efs"
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1Mi
---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: efs-provisioner-runner
rules:
  - apiGroups: [""]
    resources: ["persistentvolumes"]
    verbs: ["get", "list", "watch", "create", "delete"]
  - apiGroups: [""]
    resources: ["persistentvolumeclaims"]
    verbs: ["get", "list", "watch", "update"]
  - apiGroups: ["storage.k8s.io"]
    resources: ["storageclasses"]
    verbs: ["get", "list", "watch"]
  - apiGroups: [""]
    resources: ["events"]
    verbs: ["create", "update", "patch"]
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: run-efs-provisioner
subjects:
  - kind: ServiceAccount
    name: efs-provisioner
     # replace with namespace where provisioner is deployed
    namespace: default
roleRef:
  kind: ClusterRole
  name: efs-provisioner-runner
  apiGroup: rbac.authorization.k8s.io
---
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: leader-locking-efs-provisioner
rules:
  - apiGroups: [""]
    resources: ["endpoints"]
    verbs: ["get", "list", "watch", "create", "update", "patch"]
---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: leader-locking-efs-provisioner
subjects:
  - kind: ServiceAccount
    name: efs-provisioner
    # replace with namespace where provisioner is deployed
    namespace: default
roleRef:
  kind: Role
  name: leader-locking-efs-provisioner
  apiGroup: rbac.authorization.k8s.io
---
kind: ServiceAccount
apiVersion: v1
metadata:
  name: efs-provisioner`

**Comment by wiltonfelix** [comment]
> @tonybranfort provided the right fix.
> In summary I had to update my deployment and add ClusterRole, ClusterRoleBinding, Role, RoleBinding, ServiceAccount.
> The deployment template provided was expecting a directory on my efs volume called /persistentvolumes. I had to replace this with just / for the pod to start.
> 
> See below for configuration I used:
> 
> ```
> ---
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: efs-provisioner
> data:
>   file.system.id: xxxxxxxxxxxxxx
>   aws.region: xxxxxxxxxxxxx
>   provisioner.name: example.com/aws-efs
>   dns.name: ""
> ---
> kind: Deployment
> apiVersion: extensions/v1beta1
> metadata:
>   name: efs-provisioner
> spec:
>   replicas: 1
>   strategy:
>     type: Recreate 
>   template:
>     metadata:
>       labels:
>         app: efs-provisioner
>     spec:
>       serviceAccount: efs-provisioner
>       containers:
>         - name: efs-provisioner
>           image: quay.io/external_storage/efs-provisioner:latest
>           env:
>             - name: FILE_SYSTEM_ID
>               valueFrom:
>                 configMapKeyRef:
>                   name: efs-provisioner
>                   key: file.system.id
>             - name: AWS_REGION
>               valueFrom:
>                 configMapKeyRef:
>                   name: efs-provisioner
>                   key: aws.region
>             - name: DNS_NAME
>               valueFrom:
>                 configMapKeyRef:
>                   name: efs-provisioner
>                   key: dns.name
>                   optional: true
>             - name: PROVISIONER_NAME
>               valueFrom:
>                 configMapKeyRef:
>                   name: efs-provisioner
>                   key: provisioner.name
>           volumeMounts:
>             - name: pv-volume
>               mountPath: /persistentvolumes
>       volumes:
>         - name: pv-volume
>           nfs:
>             server: xxxxxxxxxxxxxxx.efs.xxxxxxxxxxx.amazonaws.com
>             path: /
> ---
> kind: StorageClass
> apiVersion: storage.k8s.io/v1
> metadata:
>   name: aws-efs
> provisioner: example.com/aws-efs
> ---
> kind: PersistentVolumeClaim
> apiVersion: v1
> metadata:
>   name: efs
>   annotations:
>     volume.beta.kubernetes.io/storage-class: "aws-efs"
> spec:
>   accessModes:
>     - ReadWriteMany
>   resources:
>     requests:
>       storage: 1Mi
> ---
> kind: ClusterRole
> apiVersion: rbac.authorization.k8s.io/v1
> metadata:
>   name: efs-provisioner-runner
> rules:
>   - apiGroups: [""]
>     resources: ["persistentvolumes"]
>     verbs: ["get", "list", "watch", "create", "delete"]
>   - apiGroups: [""]
>     resources: ["persistentvolumeclaims"]
>     verbs: ["get", "list", "watch", "update"]
>   - apiGroups: ["storage.k8s.io"]
>     resources: ["storageclasses"]
>     verbs: ["get", "list", "watch"]
>   - apiGroups: [""]
>     resources: ["events"]
>     verbs: ["create", "update", "patch"]
> ---
> kind: ClusterRoleBinding
> apiVersion: rbac.authorization.k8s.io/v1
> metadata:
>   name: run-efs-provisioner
> subjects:
>   - kind: ServiceAccount
>     name: efs-provisioner
>      # replace with namespace where provisioner is deployed
>     namespace: default
> roleRef:
>   kind: ClusterRole
>   name: efs-provisioner-runner
>   apiGroup: rbac.authorization.k8s.io
> ---
> kind: Role
> apiVersion: rbac.authorization.k8s.io/v1
> metadata:
>   name: leader-locking-efs-provisioner
> rules:
>   - apiGroups: [""]
>     resources: ["endpoints"]
>     verbs: ["get", "list", "watch", "create", "update", "patch"]
> ---
> kind: RoleBinding
> apiVersion: rbac.authorization.k8s.io/v1
> metadata:
>   name: leader-locking-efs-provisioner
> subjects:
>   - kind: ServiceAccount
>     name: efs-provisioner
>     # replace with namespace where provisioner is deployed
>     namespace: default
> roleRef:
>   kind: Role
>   name: leader-locking-efs-provisioner
>   apiGroup: rbac.authorization.k8s.io
> ---
> kind: ServiceAccount
> apiVersion: v1
> metadata:
>   name: efs-provisioner
> ```

@newalp this working to me, thank you!

**Comment by sdarwin** [comment]
I encountered the same error "waiting for a volume to be created, either by external provisioner "example.com/aws-efs" or manually created by system administrator" .  The solution, from the README file https://github.com/kubernetes-incubator/external-storage/tree/master/aws/efs

"""
```yaml
      volumes:
        - name: pv-volume
          nfs:
            server: fs-47a2c22e.efs.us-west-2.amazonaws.com
            path: /persistentvolumes
```
You will need to create the directory you use for `path:` on your EFS file system first or the efs-provisioner pod will fail to start.
"""

In other words, mount the NFS directory on /mnt (for example), and run mkdir /mnt/persistentvolumes .  Or change the path in deployment.yaml to / .

**Comment by chrisvdb** [comment]
For me it was solved by adding the service accounts as described in [https://github.com/kubernetes-incubator/external-storage/issues/953](https://github.com/kubernetes-incubator/external-storage/issues/953).

**Comment by debu99** [comment]
in my case, the selector part should be removed

**Comment by tjg184** [comment]
My security groups were not correctly defined on the EFS file system, which prevented the mount.  I was able to confirm this, by starting an EC2 instance (separate from the worker nodes) and tried to mount the EFS volume on the command line.    

Once I fixed security groups, it created fine with either `path: /` or `path: /persistentvolumes`

**Comment by nagapavan** [comment]
Closing this old issue. Recent comments may not be related to the actual problem I had.

**Comment by edo-lynx** [comment]
> @tonybranfort I was using Service-acccount, rbac and cluster-role binding to the provisioner. There was a bit of challenge in case of cluster-role binding case, I had to fetch existing role binding entry and patch it for each new `namespace`.
> Still ran into the issue. :(

can you please explain how you did it?

**Comment by xrow** [comment]
Hi,

I had the same issue my PVS was named "data-nfs-nfs-server-provisioner-0, but my pv was named "data-nfs-server-provisioner-0".

I was installing with helm 3 on openshift 3.11.

I think the trick is to let have the PCV the correct name "data-nfs-server-provisioner-0".

fullnameOverride: "" didn`t work either. I believe the charts needs some minor tweak.

**Comment by Omprakash50** [comment]
please help my efs-provisioner is in pending

**Comment by edo-lynx** [comment]
@Omprakash50  can you post your efs-provisioner-runner clusterrole?

**Comment by Omprakash50** [comment]
> @Omprakash50 can you post your efs-provisioner-runner clusterrole?

#efs-provisioner

kind: Deployment
apiVersion: apps/v1
metadata:
  name: efs-provisioner
spec:
  selector:
    matchLabels:
      app: efs-provisioner
  replicas: 1
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: efs-provisioner
    spec:
      containers:
        - name: efs-provisioner
          image: quay.io/external_storage/efs-provisioner:v0.1.0
          env:
            - name: FILE_SYSTEM_ID
              value: xxxxxxxxxx
            - name: AWS_REGION
              value: ap-south-1
            - name: PROVISIONER_NAME
              value: efs-storage
          volumeMounts:
            - name: pv-volume
              mountPath: /persistentvolumes
      volumes:
        - name: pv-volume
          nfs:
            server: xxxxxxxxxx.efs.ap-south-1.amazonaws.com
            path: /

**Comment by edo-lynx** [comment]
@Omprakash50  can you please also post your 'role' 
name- leader-locking-efs-provisioner
